source/pele/xy_model_online.h

```diff
@@ -0,0 +1,236 @@
+        nvertices_ = 0;
+        nedges_ = 0;
+    }
+    
+    void print(std::ostream& stm)
+    {
+        stm << "number of vertices in graph: " << nvertices() << "\n";
+        stm << "number of edges in graph: " << nedges() << "\n";
+        stm << "adjacency list:\n";
+        for (unsigned int i = 0; i < nvertices_; ++i) {
+            stm << i << " ==>> ";
+            std::shared_ptr<adjacency_list_edgenode> p = edges.at(i);
+            unsigned int tmp = 0;
+            while (p != NULL) {
+                if (tmp++) {
+                    stm << " --> ";
+                }
+                stm << p->y;
+                p = p->next;
+            }
+            stm << "\n";
+        }
+    }
+    
+    unsigned int nvertices() const
+    {
+        return nvertices_;
+    }
+    
+    unsigned int nedges() const
+    {
+        return nedges_;
+    }
+    
+    bool directed() const
+    {
+        return directed_;
+    }
+    
+    unsigned int maxv() const
+    {
+        return edges.size();
+    }
+    
+    std::shared_ptr<adjacency_list_edgenode> get_edges(const unsigned int v) const
+    {
+        return edges.at(v);
+    }
+};
+
+/**
+ * XY model implementation where the energy is computed by summing each
+ * linked list in the adjacency list vector in tun. 
+ */
+class XYModelOnline : public BasePotentialOnline {
+    adjacency_list_graph m_topology;
+public:
+    XYModelOnline(const size_t nr_spins, pele::Array<size_t> head, pele::Array<size_t> tail)
+        : BasePotentialOnline(nr_spins),
+          m_topology(nr_spins, head, tail)
+    {}
+        
+    double get_energy(Array<double> x, const size_t batch_number)
+    {
+        /**
+         * Return ith term of the potential energy, i.e., the sum of 
+         * -cos(x[i] - x[j]) for all j which are linked to i
+         */
+        if (x.size() != m_topology.nvertices()) {
+            throw std::runtime_error("XYModelOnline: x.size() != nr vertices in graph");
+        }
+        double energy = 0;
+        std::shared_ptr<adjacency_list_edgenode> en = m_topology.get_edges(batch_number);
+        while (en) {
+            energy += -std::cos(x[batch_number] - x[en->y]);
+            en = en->next;
+        }
+        return energy;
+    }
+    
+    double get_energy_gradient_batch(Array<double> x, const size_t batch_number,
+        Array<double> ograd)
+    {
+        /**
+         * Return full potential energy. Compute
+         * gradient of ith term.
+         */
+        if (x.size() != m_topology.nvertices() || ograd.size() != x.size()) {
+            throw std::runtime_error("XYModelOnline: illegal input");
+        }
+        ograd.assign(0);
+        std::shared_ptr<adjacency_list_edgenode> en = m_topology.get_edges(batch_number);
+        while (en) {
+            const double tmp = std::sin(x[batch_number] - x[en->y]);
+            for (size_t k = 0; k < x.size(); ++k) {
+                ograd[k] += ((k == batch_number) - (k == en->y)) * tmp;
```

Comment:
I am not sure I understand this. Do we need to loop over x.size() when we need to perform the operation only on 2 elements? Can't we just do
`if (batch_number != en->y){
    ograd[batch_number] += tmp;
    ograd[en->y] -= tmp;}`
I don't think we even need the conditional statement, I just added it to reproduce what the for loop is doing.